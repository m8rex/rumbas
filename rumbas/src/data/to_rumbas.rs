use crate::data::question_part::question_part::QuestionPart;
use crate::data::translatable::ContentAreaTranslatableString;
use crate::data::translatable::EmbracedJMETranslatableString;
use crate::data::translatable::JMETranslatableString;
use crate::data::translatable::TranslatableString;
use numbas::defaults::DEFAULTS;
use numbas::jme::{ContentAreaString, EmbracedJMEString, JMEString};
use std::convert::TryInto;

pub trait ToRumbas<RumbasType>: Clone {
    fn to_rumbas(&self) -> RumbasType;
}

macro_rules! impl_to_rumbas {
    ($($type: ty$([$($gen: tt), *])?), *) => {
        $(
        impl$(< $($gen : Clone),* >)? ToRumbas<$type> for $type {
            fn to_rumbas(&self) -> $type {
                self.clone()
            }
        }
        )*
    };
}

impl_to_rumbas!(bool, f64, usize, [f64; 2]);
impl_to_rumbas!(numbas::exam::Primitive);
impl_to_rumbas!(numbas::jme::JMEString);
impl_to_rumbas!(numbas::jme::EmbracedJMEString);
impl_to_rumbas!(numbas::jme::ContentAreaString);

impl<T, O: ToRumbas<T>> ToRumbas<Vec<T>> for Vec<O> {
    fn to_rumbas(&self) -> Vec<T> {
        self.iter().map(|item| item.to_rumbas()).collect()
    }
}

impl<T, O: ToRumbas<T>> ToRumbas<Option<T>> for Option<O> {
    fn to_rumbas(&self) -> Option<T> {
        self.clone().map(|item| item.to_rumbas())
    }
}

impl ToRumbas<TranslatableString> for String {
    fn to_rumbas(&self) -> TranslatableString {
        self.clone().into()
    }
}

impl ToRumbas<JMETranslatableString> for JMEString {
    fn to_rumbas(&self) -> JMETranslatableString {
        self.clone().into()
    }
}

impl ToRumbas<EmbracedJMETranslatableString> for EmbracedJMEString {
    fn to_rumbas(&self) -> EmbracedJMETranslatableString {
        self.clone().into()
    }
}

impl ToRumbas<ContentAreaTranslatableString> for ContentAreaString {
    fn to_rumbas(&self) -> ContentAreaTranslatableString {
        self.clone().into()
    }
}

pub fn extract_part_common_marks(
    pd: &numbas::exam::ExamQuestionPartSharedData,
) -> numbas::exam::Primitive {
    pd.marks
        .clone()
        .unwrap_or(numbas::exam::Primitive::Natural(DEFAULTS.part_common_marks))
}

pub fn extract_part_common_prompt(
    pd: &numbas::exam::ExamQuestionPartSharedData,
) -> ContentAreaTranslatableString {
    pd.prompt
        .clone()
        .unwrap_or_else(|| "".to_string().try_into().unwrap())
        .to_rumbas()
}

pub fn extract_part_common_use_custom_name(pd: &numbas::exam::ExamQuestionPartSharedData) -> bool {
    pd.use_custom_name
        .unwrap_or(DEFAULTS.part_common_use_custom_name)
}

pub fn extract_part_common_custom_name(pd: &numbas::exam::ExamQuestionPartSharedData) -> String {
    pd.custom_name.clone().unwrap_or_default()
}

pub fn extract_part_common_steps_penalty(pd: &numbas::exam::ExamQuestionPartSharedData) -> usize {
    pd.steps_penalty
        .unwrap_or(DEFAULTS.part_common_steps_penalty)
}

pub fn extract_part_common_enable_minimum_marks(
    pd: &numbas::exam::ExamQuestionPartSharedData,
) -> bool {
    pd.enable_minimum_marks
        .unwrap_or(DEFAULTS.part_common_enable_minimum_marks)
}

pub fn extract_part_common_minimum_marks(pd: &numbas::exam::ExamQuestionPartSharedData) -> usize {
    pd.minimum_marks
        .unwrap_or(DEFAULTS.part_common_minimum_marks)
}

pub fn extract_part_common_show_correct_answer(
    pd: &numbas::exam::ExamQuestionPartSharedData,
) -> bool {
    pd.show_correct_answer
}

pub fn extract_part_common_show_feedback_icon(
    pd: &numbas::exam::ExamQuestionPartSharedData,
) -> bool {
    pd.show_feedback_icon
        .unwrap_or(DEFAULTS.part_common_show_feedback_icon)
}

pub fn extract_part_common_adaptive_marking_penalty(
    pd: &numbas::exam::ExamQuestionPartSharedData,
) -> usize {
    pd.adaptive_marking_penalty
        .unwrap_or(DEFAULTS.part_common_adaptive_marking_penalty)
}

pub fn extract_part_common_extend_base_marking_algorithm(
    pd: &numbas::exam::ExamQuestionPartSharedData,
) -> bool {
    pd.extend_base_marking_algorithm
        .unwrap_or(DEFAULTS.part_common_extend_base_marking_algorithm)
}

pub fn extract_part_common_steps(
    pd: &numbas::exam::ExamQuestionPartSharedData,
) -> Vec<QuestionPart> {
    pd.steps
        .clone()
        .unwrap_or_default()
        .into_iter()
        .map(|s| s.to_rumbas())
        .collect()
}
