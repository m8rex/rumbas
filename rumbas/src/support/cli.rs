use clap::{Parser, Subcommand};
use clap_complete::Shell;

/// The rumbas cli
#[derive(Debug, Parser)] // requires `derive` feature
#[clap[author, version, about, long_about = None]]
pub struct Cli {
    #[clap(subcommand)]
    pub command: Command,

    /// Sets the level of verbosity
    ///
    /// -v    sets the level to ERROR
    ///
    /// -vv   sets the level to WARN
    ///
    /// -vvv   sets the level to INFO
    ///
    /// -vvvv   sets the level to DEBUG
    ///
    /// The default is -vvv
    #[clap(short, action = clap::ArgAction::Count)]
    pub verbose: u8,
    /// Enables quiet mode. Nothing is logged. This has precedence over the verbose option.
    #[clap(short)]
    pub quiet: bool,
}

#[derive(Debug, Subcommand)]
pub enum Command {
    /// Compile a rumbas exam (or question)
    ///
    /// You can pass a path to a folder to compile all files in the folder.
    #[clap(arg_required_else_help = true)]
    Compile {
        /// The path to the exam or question file to compile.
        ///
        /// If a folder within the questions or exams folder is used, all questions/exams in that folder will be compiled.
        ///
        /// It is possible to specify multiple paths to folder/files.
        #[clap(required = true, value_parser)]
        exam_or_question_paths: Vec<String>,
        /// Include the files necessary to make a SCORM package
        #[clap(value_parser, long, short)]
        scorm: bool,
        /// Create a zip file instead of a directory
        #[clap(value_parser, long, short)]
        zip: bool,
        /// Don't perform minification on the created js in the exam. Useful if you don't have uglifyjs or want to debug something.
        #[clap(value_parser, long)]
        no_minification: bool,
    },
    /// Check a rumbas exam (or question)
    ///
    /// You can pass a path to a folder to check all files in the folder.
    #[clap(arg_required_else_help = true)]
    Check {
        /// The path to the exam or question file to check.
        ///
        /// If a folder within the questions or exams folder is used, all questions/exams in that folder will be checked.
        ///
        /// It is possible to specify multiple paths to folder/files.
        #[clap(required = true, value_parser)]
        exam_or_question_paths: Vec<String>,
    },
    /// Watch a path
    #[clap(arg_required_else_help = true)]
    Watch {
        /// The path to watch
        path: String,
        /// Only check exams and questions that change due to file changes, but don't compile them with numbas.
        #[clap(short)]
        only_check: bool,
    },
    /// Format a rumbas exam (or question).
    ///
    /// You can pass a path to a folder to format all files in the folder.
    #[clap(arg_required_else_help = true)]
    Fmt {
        /// The path to the exam or question file to format.
        ///
        /// If a folder within the questions or exams folder is used, all questions/exams in that folder will be formatted.
        ///
        /// It is possible to specify multiple paths to folder/files.
        #[clap(required = true, value_parser)]
        exam_or_question_paths: Vec<String>,
    },
    /// Import a numbas .exam file
    ///    
    /// Resources have to be manually placed in the resources folder
    #[clap(arg_required_else_help = true)]
    Import {
        /// The path to the numbas .exam file
        #[clap(value_parser)]
        exam_path: String,
        /// Tells rumbas that this is the exam file of a numbas question instead of of a numbas exam.
        #[clap(short)]
        question: bool,
    },
    /// Initialize a rumbas project in this folder
    Init {
        /// Whether the defaults should be of summative nature (instead of formative)
        #[clap(value_parser, long)]
        summative: bool,
    },
    /// Update the repository to the next rumbas version
    UpdateRepo,
    /// Export a rumbas exam as one yaml.
    /// All default files and templating is resolved.
    ///
    /// Can be useful to debug exams / questions that don't behave as expected
    Export {
        /// The path to the exams to export
        ///
        /// If a folder within the questions or exams folder is used, all questions/exams in that folder will be exported.
        ///
        /// It is possible to specify multiple paths to folder/files.
        #[clap(required = true, value_parser)]
        exam_or_question_paths: Vec<String>,
    },
    /// Creates files with the json schemas (beta).
    /// See https://github.com/m8rex/rumbas-examples/tree/main/.vscode for usage instructions
    Schema,
    /// Generate shell completion file. Placing this file at the right location and reloading your
    /// shell, will give you shell completion for rumbas.
    ///
    /// Example for bash: rumbas generate-shell-completion bash | sudo tee /usr/share/bash-completion/completions/rumbas
    #[clap(arg_required_else_help = true)]
    GenerateShellCompletion {
        /// The shell for which the shell competion should be generated
        shell: Shell,
    },
    /// Generates a folder structure that can be hosted and used as an 'editor' in the numbas lti provider
    ///
    /// Only exams are compiled.
    #[clap(arg_required_else_help = true)]
    EditorOutput {
        /// The path to the folder where the output should be generated.
        output_path: String,
        /// The url prefix for all editor api calls
        url_prefix: String,
    },
}

impl Command {
    fn can_execute_in_old_version(&self) -> bool {
        matches!(
            self,
            Self::UpdateRepo
                | Self::Init { summative: _ }
                | Self::GenerateShellCompletion { shell: _ }
        )
    }
}
