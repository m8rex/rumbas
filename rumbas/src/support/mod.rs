#[macro_use]
pub mod optional_overwrite;
#[macro_use]
pub mod to_numbas;
#[macro_use]
pub mod to_rumbas;
pub mod default;
pub mod template;
pub mod yaml;
