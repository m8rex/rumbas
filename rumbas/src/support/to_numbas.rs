use crate::support::optional_overwrite::*;

pub type NumbasResult<T> = Result<T, RumbasCheckResult>;

pub trait ToNumbas<NumbasType>: Clone + RumbasCheck {
    /// Method that safely converts a type to another (probably numbas) type
    fn to_numbas_safe(&self, locale: &str) -> NumbasResult<NumbasType> {
        let check = self.check(locale);
        if check.is_empty() {
            Ok(self.to_numbas(locale))
        } else {
            Err(check)
        }
    }
    /// Method that converts a type to another type
    /// This method assumes that it is called by a function that is initially called from `to_numbas_safe`
    fn to_numbas(&self, locale: &str) -> NumbasType;
    fn to_numbas_with_name(&self, locale: &str, _name: String) -> NumbasType {
        self.to_numbas(locale)
    }
}

impl_to_numbas!(String, bool, f64, usize);
impl_to_numbas!(numbas::jme::JMEString);
impl_to_numbas!(numbas::jme::EmbracedJMEString);
impl_to_numbas!(numbas::jme::ContentAreaString);

impl<S, O: ToNumbas<S>> ToNumbas<Vec<S>> for Vec<O> {
    fn to_numbas(&self, locale: &str) -> Vec<S> {
        let mut v = Vec::new();
        for item in self.iter() {
            v.push(item.to_numbas(locale));
        }
        v
    }
}

impl<K: Clone + std::hash::Hash + std::cmp::Eq, S, O: ToNumbas<S>>
    ToNumbas<std::collections::HashMap<K, S>> for std::collections::HashMap<K, O>
where
    std::collections::HashMap<K, O>: RumbasCheck,
{
    fn to_numbas(&self, locale: &str) -> std::collections::HashMap<K, S> {
        self.iter()
            .map(|(k, v)| (k.to_owned(), v.to_numbas(locale)))
            .collect()
    }
}

impl<AA, A: ToNumbas<AA>, BB, B: ToNumbas<BB>> ToNumbas<(AA, BB)> for (A, B)
where
    (A, B): RumbasCheck,
{
    fn to_numbas(&self, locale: &str) -> (AA, BB) {
        (self.0.to_numbas(locale), self.1.to_numbas(locale))
    }
}

impl<AA, A: ToNumbas<AA>> ToNumbas<[AA; 2]> for [A; 2]
where
    [A; 2]: RumbasCheck,
{
    fn to_numbas(&self, locale: &str) -> [AA; 2] {
        [self[0].to_numbas(locale), self[1].to_numbas(locale)]
    }
}

impl ToNumbas<numbas::exam::SafeFloat> for f64 {
    fn to_numbas(&self, _locale: &str) -> numbas::exam::SafeFloat {
        (*self).into()
    }
}

impl ToNumbas<numbas::exam::SafeNatural> for usize {
    fn to_numbas(&self, _locale: &str) -> numbas::exam::SafeNatural {
        (*self).into()
    }
}

macro_rules! impl_to_numbas {
    ($($type: ty), *) => {
        $(
        impl ToNumbas<$type> for $type {
            fn to_numbas(&self, _locale: &str) -> $type {
                self.clone()
            }
        }
        )*
    };
}

pub(crate) use impl_to_numbas;
