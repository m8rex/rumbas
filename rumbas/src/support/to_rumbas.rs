use crate::question::part::question_part::QuestionPart;
use crate::support::file_reference::{FileString, JMEFileString};
use crate::support::translatable::ContentAreaTranslatableString;
use crate::support::translatable::EmbracedJMETranslatableString;
use crate::support::translatable::JMETranslatableString;
use crate::support::translatable::TranslatableString;
use numbas::jme::{ContentAreaString, EmbracedJMEString, JMEString};
use numbas::support::primitive::Number;
use std::convert::TryInto;

pub trait ToRumbas<RumbasType>: Clone {
    fn to_rumbas(&self) -> RumbasType;
}

impl_to_rumbas!(String, bool, f64, usize, [f64; 2]);
impl_to_rumbas!(numbas::support::primitive::Number);
impl_to_rumbas!(numbas::jme::JMEString);
impl_to_rumbas!(numbas::jme::EmbracedJMEString);
impl_to_rumbas!(numbas::jme::ContentAreaString);

impl<T, O: ToRumbas<T>> ToRumbas<Vec<T>> for Vec<O> {
    fn to_rumbas(&self) -> Vec<T> {
        self.iter().map(|item| item.to_rumbas()).collect()
    }
}

impl<K: Clone + std::hash::Hash + std::cmp::Eq, S, O: ToRumbas<S>>
    ToRumbas<std::collections::HashMap<K, S>> for std::collections::HashMap<K, O>
{
    fn to_rumbas(&self) -> std::collections::HashMap<K, S> {
        self.iter()
            .map(|(k, v)| (k.to_owned(), v.to_rumbas()))
            .collect()
    }
}

impl<K, L: ToRumbas<K>, S, O: ToRumbas<S>> ToRumbas<(K, S)> for (L, O) {
    fn to_rumbas(&self) -> (K, S) {
        (self.0.to_rumbas(), self.1.to_rumbas())
    }
}

impl<T, O: ToRumbas<T>> ToRumbas<Option<T>> for Option<O> {
    fn to_rumbas(&self) -> Option<T> {
        self.clone().map(|item| item.to_rumbas())
    }
}

impl ToRumbas<FileString> for String {
    fn to_rumbas(&self) -> FileString {
        FileString::s(self)
    }
}

impl ToRumbas<JMEFileString> for String {
    fn to_rumbas(&self) -> JMEFileString {
        JMEFileString::s(self)
    }
}

impl ToRumbas<TranslatableString> for String {
    fn to_rumbas(&self) -> TranslatableString {
        self.clone().into()
    }
}

impl ToRumbas<JMETranslatableString> for JMEString {
    fn to_rumbas(&self) -> JMETranslatableString {
        self.clone().into()
    }
}

impl ToRumbas<EmbracedJMETranslatableString> for EmbracedJMEString {
    fn to_rumbas(&self) -> EmbracedJMETranslatableString {
        self.clone().into()
    }
}

impl ToRumbas<ContentAreaTranslatableString> for ContentAreaString {
    fn to_rumbas(&self) -> ContentAreaTranslatableString {
        self.clone().into()
    }
}

pub fn extract_part_common_marks(
    pd: &numbas::question::part::QuestionPartSharedData,
) -> numbas::support::primitive::Number {
    pd.marks.clone()
}

pub fn extract_part_common_prompt(
    pd: &numbas::question::part::QuestionPartSharedData,
) -> ContentAreaTranslatableString {
    pd.prompt.clone().to_rumbas()
}

pub fn extract_part_common_use_custom_name(
    pd: &numbas::question::part::QuestionPartSharedData,
) -> bool {
    pd.use_custom_name
}

pub fn extract_part_common_custom_name(
    pd: &numbas::question::part::QuestionPartSharedData,
) -> String {
    pd.custom_name.clone()
}

pub fn extract_part_common_steps_penalty(
    pd: &numbas::question::part::QuestionPartSharedData,
) -> Number {
    pd.steps_penalty.clone()
}

pub fn extract_part_common_enable_minimum_marks(
    pd: &numbas::question::part::QuestionPartSharedData,
) -> bool {
    pd.enable_minimum_marks
}

pub fn extract_part_common_minimum_marks(
    pd: &numbas::question::part::QuestionPartSharedData,
) -> usize {
    pd.minimum_marks
}

pub fn extract_part_common_show_correct_answer(
    pd: &numbas::question::part::QuestionPartSharedData,
) -> bool {
    pd.show_correct_answer
}

pub fn extract_part_common_show_feedback_icon(
    pd: &numbas::question::part::QuestionPartSharedData,
) -> bool {
    pd.show_feedback_icon
}

pub fn extract_part_common_adaptive_marking_penalty(
    pd: &numbas::question::part::QuestionPartSharedData,
) -> usize {
    pd.adaptive_marking_penalty
}

pub fn extract_part_common_extend_base_marking_algorithm(
    pd: &numbas::question::part::QuestionPartSharedData,
) -> bool {
    pd.extend_base_marking_algorithm
}

pub fn extract_part_common_steps(
    pd: &numbas::question::part::QuestionPartSharedData,
) -> Vec<QuestionPart> {
    pd.steps.to_rumbas()
}

/// Macro used to create a question part type for numbas
/// Usage: create_question_part! { PartName with &self.part_data => { field1: val1, field2, val2 }  }
macro_rules! create_question_part {
    (
        $type: ident with $part_data: expr => {
            $(
                $field: ident$(: $val: expr)?
            ),*
        }
    ) => {
        {
            let part_data = $part_data;
            let custom_marking_algorithm_notes: JMENotes = part_data
                .custom_marking_algorithm
                .clone()
                .to_rumbas();
            $type {
                // Default section
                marks: extract_part_common_marks(&part_data).to_rumbas(),
                prompt: extract_part_common_prompt(&part_data),
                use_custom_name: extract_part_common_use_custom_name(&part_data).to_rumbas(),
                custom_name: extract_part_common_custom_name(&part_data).to_rumbas(),
                steps_penalty: extract_part_common_steps_penalty(&part_data).to_rumbas(),
                enable_minimum_marks: extract_part_common_enable_minimum_marks(&part_data)
                    .to_rumbas(),
                minimum_marks: extract_part_common_minimum_marks(&part_data).to_rumbas(),
                show_correct_answer: extract_part_common_show_correct_answer(&part_data)
                    .to_rumbas(),
                show_feedback_icon: extract_part_common_show_feedback_icon(&part_data).to_rumbas(),
                variable_replacement_strategy: part_data.variable_replacement_strategy.to_rumbas(),
                adaptive_marking_penalty: extract_part_common_adaptive_marking_penalty(&part_data)
                    .to_rumbas(),
                custom_marking_algorithm_notes,
                extend_base_marking_algorithm: extract_part_common_extend_base_marking_algorithm(
                    &part_data,
                )
                .to_rumbas(),
                steps: extract_part_common_steps(&part_data),
                $(
                    $field$(: $val)?
                ),*
            }
        }
    }
}
pub(crate) use create_question_part;

macro_rules! impl_to_rumbas {
    ($($type: ty$([$($gen: tt), *])?), *) => {
        $(
        impl$(< $($gen : Clone),* >)? ToRumbas<$type> for $type {
            fn to_rumbas(&self) -> $type {
                self.clone()
            }
        }
        )*
    };
}
pub(crate) use impl_to_rumbas;
