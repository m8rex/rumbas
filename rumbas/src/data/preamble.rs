use crate::data::file_reference::FileString;
use crate::data::optional_overwrite::{Noneable, OptionalOverwrite};
use crate::data::to_numbas::{NumbasResult, ToNumbas};
use serde::{Deserialize, Serialize};

// TODO: maybe translatable? So different text's can be set for different languages?
optional_overwrite! {
    Preamble,
    js: FileString,
    css: FileString
}

impl ToNumbas for Preamble {
    type NumbasType = numbas::exam::Preamble;
    fn to_numbas(&self, _locale: &String) -> NumbasResult<numbas::exam::Preamble> {
        let empty_fields = self.empty_fields();
        if empty_fields.is_empty() {
            Ok(numbas::exam::Preamble::new(
                self.js.clone().unwrap().get_content(),
                self.css.clone().unwrap().get_content(),
            ))
        } else {
            Err(empty_fields)
        }
    }
}
