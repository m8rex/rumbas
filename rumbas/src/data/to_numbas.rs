use crate::data::optional_overwrite::{Noneable, OptionalOverwrite};
use crate::data::template::{Value, ValueType};

pub type NumbasResult<T> = Result<T, Vec<String>>;

pub trait ToNumbas: Clone {
    type NumbasType;
    fn to_numbas(&self, locale: &String) -> NumbasResult<Self::NumbasType>;
    fn to_numbas_with_name(
        &self,
        locale: &String,
        _name: String,
    ) -> NumbasResult<Self::NumbasType> {
        self.to_numbas(&locale)
    }
}

impl<T: ToNumbas + OptionalOverwrite> ToNumbas for Value<T> {
    type NumbasType = <T as ToNumbas>::NumbasType;
    fn to_numbas(&self, locale: &String) -> NumbasResult<Self::NumbasType> {
        match &self.0 {
            Some(ValueType::Normal(val)) => {
                let empty_fields = val.empty_fields();
                if empty_fields.is_empty() {
                    Ok(val.to_numbas(&locale).unwrap())
                } else {
                    Err(empty_fields)
                }
            }
            Some(ValueType::Template(ts)) => Err(vec![ts.yaml()]),
            None => Err(vec!["".to_string()]),
        }
    }
}

impl<T: ToNumbas + OptionalOverwrite> ToNumbas for Noneable<T> {
    type NumbasType = Option<<T as ToNumbas>::NumbasType>;
    fn to_numbas(&self, locale: &String) -> NumbasResult<Self::NumbasType> {
        match self {
            Noneable::NotNone(val) => {
                let empty_fields = val.empty_fields();
                if empty_fields.is_empty() {
                    Ok(Some(val.clone().to_numbas(&locale).unwrap()))
                } else {
                    Err(empty_fields)
                }
            }
            _ => Ok(None),
        }
    }
}

macro_rules! impl_to_numbas {
    ($($type: ty), *) => {
        $(
        impl ToNumbas for $type {
            type NumbasType = $type;
            fn to_numbas(&self, _locale: &String) -> NumbasResult<Self::NumbasType> {
                Ok(self.clone())
            }
        }
        )*
    };
}

impl_to_numbas!(String, bool, f64, usize, [f64; 2]);

impl<O: ToNumbas> ToNumbas for Vec<O> {
    type NumbasType = Vec<O::NumbasType>;
    fn to_numbas(&self, locale: &String) -> NumbasResult<Self::NumbasType> {
        let mut v = Vec::new();
        for item in self.into_iter() {
            v.push(item.to_numbas(&locale)?);
        }
        Ok(v)
    }
}
