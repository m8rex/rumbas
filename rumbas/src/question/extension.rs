use crate::support::to_numbas::ToNumbas;
use crate::support::to_rumbas::*;
use comparable::Comparable;
use rumbas_support::preamble::*;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use structdoc::StructDoc;

macro_rules! extensions {
    (
        $(
            $(#[$inner:meta])*
            $name: ident: $path: literal
        ),+
    ) => {
            #[derive(Input, Overwrite, RumbasCheck, Examples, StructDoc)]
            #[input(name = "ExtensionsInput")]
            #[derive(Serialize, Deserialize, Comparable, Debug, Clone, JsonSchema, PartialEq, Eq)]
            /// Specify which extensions should be enabled
            pub struct Extensions {
                $(
                    $(
                        #[$inner]
                    )*
                    pub $name: bool
                ),*
            }

        impl ToNumbas<Vec<String>> for Extensions {
            type ToNumbasHelper = ();
            fn to_numbas(&self, locale: &str, _data: &Self::ToNumbasHelper) -> Vec<String> {
                let mut extensions = Vec::new();
                $(
                    if self.$name.to_numbas(locale, &()) {
                        extensions.push($path.to_string()); //TODO: Enum in numbas crate?
                    }
                )*
                extensions
            }
        }

        impl Extensions {
            pub fn from(v: &[String]) -> Self {
                Extensions {
                    $(
                        $name: v.contains(&$path.to_string()).to_rumbas()
                    ),*
                }
            }

            pub fn combine(e: Extensions, f: Extensions) -> Extensions {
                Extensions {
                    $(
                    $name: e.$name || f.$name
                    ),*
                }
            }

            pub fn to_paths(&self) -> Vec<String> {
                let numbas_path = std::env::var(crate::NUMBAS_FOLDER_ENV)
                    .expect(&format!("{} to be set", crate::NUMBAS_FOLDER_ENV)[..]);
                let mut paths = Vec::new();
                $(
                    if self.$name {
                        paths.push($path);
                    }
                )*
                paths
                    .into_iter()
                    .map(|s| format!("{}/extensions/{}", numbas_path, s))
                    .collect()
            }
        }

        impl Default for Extensions {
            fn default() -> Extensions {
                Extensions {
                    $(
                        $name: false
                    ),*
                }
            }
        }

    }
}

//fixme vis.js extension? (where to find?)
extensions! {
    chemistry: "chemistry",
    download_text_file: "download-text-file",
    /// The Eukleides extension provides functions to embed diagrams created using the
    /// Eukleides language.
    /// https://numbas-editor.readthedocs.io/en/latest/extensions/first-party.html#eukleides
    eukleides: "eukleides",
    /// The GeoGebra extension provides functions to embed GeoGebra worksheets in a question.
    /// https://numbas-editor.readthedocs.io/en/latest/extensions/first-party.html#geogebra
    geogebra: "geogebra",
    /// This extension provides some functions for working with and drawing graphs (networks of vertices joined by edges) in Numbas.
    /// https://numbas-editor.readthedocs.io/en/latest/extensions/first-party.html#graph-theory
    graphs: "graphs",
    /// The JSXGraph extension provides functions to create and manipulate interactive diagrams with the JSXGraph library.
    /// https://numbas-editor.readthedocs.io/en/latest/extensions/first-party.html#eukleides
    jsx_graph: "jsxgraph",
    linear_algebra: "linear-algebra",
    /// This extension provides a new data type and some functions to deal with linear codes.
    /// https://numbas-editor.readthedocs.io/en/latest/extensions/first-party.html#linear-codes
    linear_codes: "codewords",
    optimisation: "optimisation",
    permutations: "permutations",
    /// This extension provides a new data type and some functions to deal with rational polynomials.
    /// https://numbas-editor.readthedocs.io/en/latest/extensions/first-party.html#polynomials
    polynomials: "polynomials",
    /// This extension wraps the js-quantities library to provide a “quantity with units” data type to Numbas.
    /// https://numbas-editor.readthedocs.io/en/latest/extensions/first-party.html#quantities
    quantities: "quantities",
    /// The “random person” extension provides a collection of functions to generate random people, for use in word problems.
    /// https://numbas-editor.readthedocs.io/en/latest/extensions/first-party.html#random-person
    random_person: "random_person",
    /// The statistical functions extension provides many new functions for generating samples from random distributions, and calculating statistics.
    /// https://numbas-editor.readthedocs.io/en/latest/extensions/first-party.html#statistical-functions
    stats: "stats",
    sqlite: "sqlite",
    text: "text",
    written_number: "written-number"

    // TODO: programming extension
}

impl ToRumbas<Extensions> for Vec<String> {
    fn to_rumbas(&self) -> Extensions {
        Extensions::from(self)
    }
}
