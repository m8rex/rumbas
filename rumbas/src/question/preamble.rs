use crate::support::file_reference::FileString;
use crate::support::optional_overwrite::*;
use crate::support::template::{Value, ValueType};
use crate::support::to_numbas::ToNumbas;
use crate::support::to_rumbas::ToRumbas;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

// TODO: maybe translatable? So different text's can be set for different languages?
optional_overwrite! {
    pub struct Preamble {
        /// The JavaScript to add to the outputfiles
        js: FileString,
        /// The CSS to add to the outputfiles
        css: FileString
    }
}

impl ToNumbas<numbas::exam::Preamble> for Preamble {
    fn to_numbas(&self, locale: &str) -> numbas::exam::Preamble {
        numbas::exam::Preamble {
            js: self.js.to_numbas(locale),
            css: self.css.to_numbas(locale),
        }
    }
}

impl ToRumbas<Preamble> for numbas::exam::Preamble {
    fn to_rumbas(&self) -> Preamble {
        Preamble {
            js: self.js.to_rumbas(),
            css: self.css.to_rumbas(),
        }
    }
}
