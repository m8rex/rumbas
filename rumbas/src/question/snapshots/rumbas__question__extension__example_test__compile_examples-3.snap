---
source: rumbas/src/question/extension.rs
expression: "&example"

---
chemistry: ~
download_text_file: ~
eukleides: ~
geogebra: ~
graphs: ~
jsx_graph: ~
linear_algebra: ~
linear_codes: ~
optimisation: ~
permutations: ~
polynomials: ~
quantities: ~
random_person: ~
stats: ~
sqlite: ~
written_number: ~

